File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.cc
//BFS로 영역 구하기 문제처럼 하면 될듯
#include<iostream>
#include<queue>
#include<vector>
#include<string>
#include<algorithm>
using namespace std;

#define MAX 26
bool vis[MAX][MAX] = { 0, };
string board[MAX];
int N;
queue<pair<int, int>> q;
vector<int> w;

void BFS(int a, int b) {
	int dx[] = { 1,0,-1,0 }, dy[] = { 0,-1,0,1 };
	int cnt = 0;	

	q.push({ a,b });
	vis[a][b] = 1;
	cnt++;
	while (!q.empty()) {
		int x = q.front().first;
		int y = q.front().second;
		q.pop();
		
		for (auto i = 0; i < 4; ++i) {
			int nx = x + dx[i];
			int ny = y + dy[i];
			//if (nx < 0 || ny < 0 || nx >= N || ny >= N) continue;
			if (nx >= 0 && ny >= 0 && nx < N && ny < N && vis[nx][ny]==false && board[nx][ny] == '1') {
				q.push({ nx,ny });
				vis[nx][ny] = true;
				//영역의 넓이 +1
				cnt++;
			}
		}
	}
	w.push_back(cnt);
}

int main(void) {
	cin.tie(NULL)->sync_with_stdio(false);


	cin >> N;
	for (auto i = 0; i < N; ++i) {
			cin >> board[i];
	}

	for (auto i = 0; i < N; ++i) {
		for (auto j = 0; j < N; ++j) {
			if (board[i][j] == '1' && vis[i][j]==false) {
				BFS(i,j);
			}
		}
	}


	sort(w.begin(), w.end());

	cout << w.size() << '\n';
	for (auto num : w) {
		cout << num << '\n';
	}
	return 0;
}

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.util.*;
import java.io.*;

public class Main {
    static int N;
    static int[][] board;
    static boolean[][] vis;
    static int[] dx = {0, 1, 0, -1};
    static int[] dy = {1, 0, -1, 0};
    static StringBuilder sb = new StringBuilder();
    static List<Integer> result = new ArrayList<>();

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        board = new int[N][N];
        vis = new boolean[N][N];

        for (int i = 0; i < N; i++) {
            String input = br.readLine();
            for (int j = 0; j < N; j++) {
                board[i][j] = input.charAt(j) - '0';
                if (board[i][j] == 1) board[i][j] = -1;
            }
        }

        int cnt = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N; j++) {
                if (board[i][j] == -1 && vis[i][j] == false) {
                    cnt++;
                    BFS(i, j, cnt);
                }
            }
        }
        sb.append(cnt).append("\n");
        Collections.sort(result);
        for (int n : result) {
            sb.append(n).append("\n");
        }

        System.out.println(sb.toString());
    }

    static void BFS(int x, int y, int num) {
        Queue<int[]> q = new ArrayDeque<>();
        vis[x][y] = true;
        board[x][y] = num;
        q.offer(new int[]{x, y});

        int count = 1;
        while (!q.isEmpty()) {
            int[] cur = q.poll();
            int cx = cur[0];
            int cy = cur[1];

            for (int i = 0; i < 4; i++) {
                int nx = cx + dx[i];
                int ny = cy + dy[i];

                if (nx < 0 || ny < 0 || nx >= N || ny >= N) continue;
                if (vis[nx][ny]) continue;
                if (board[nx][ny] == -1) {
                    vis[nx][ny] = true;
                    board[nx][ny] = num;
                    count++;
                    q.offer(new int[]{nx, ny});
                }
            }
        }
        result.add(count);
    }
}
